src/Compilers/CSharp/Portable/Symbols/Source/SourceNamedTypeSymbol.cs

```diff
@@ -393,44 +377,191 @@ private static SyntaxList<TypeParameterConstraintClauseSyntax> GetConstraintClau
+
+                if (constraint.Constraints != mergedKind || mergedConstraintTypes != null)
+                {
+                    Debug.Assert((constraint.Constraints & TypeParameterConstraintKind.AllNonNullableKinds) == (mergedKind & TypeParameterConstraintKind.AllNonNullableKinds));
+                    Debug.Assert((mergedKind & TypeParameterConstraintKind.ObliviousNullabilityIfReferenceType) == 0 ||
+                                 (constraint.Constraints & TypeParameterConstraintKind.ObliviousNullabilityIfReferenceType) != 0);
+                    Debug.Assert((constraint.Constraints & TypeParameterConstraintKind.AllReferenceTypeKinds) == (mergedKind & TypeParameterConstraintKind.AllReferenceTypeKinds) ||
+                                 (constraint.Constraints & TypeParameterConstraintKind.AllReferenceTypeKinds) == TypeParameterConstraintKind.ReferenceType);
+#if DEBUG
+                    if (mergedConstraintTypes != null)
+                    {
+                        Debug.Assert(originalConstraintTypes.Length == mergedConstraintTypes.Count);
+
+                        for (int j = 0; j < originalConstraintTypes.Length; j++)
+                        {
+                            Debug.Assert(TypeWithAnnotations.EqualsComparer.UnknownNullableModifierMatchesAnyComparer.Equals(originalConstraintTypes[j], mergedConstraintTypes[j]));
+                        }
+                    }
+#endif
+                    if (builder == null)
+                    {
+                        builder = ArrayBuilder<TypeParameterConstraintClause>.GetInstance(constraintClauses.Length);
+                        builder.AddRange(constraintClauses);
+                    }
+
+                    builder[i] = TypeParameterConstraintClause.Create(mergedKind,
+                                                                      mergedConstraintTypes?.ToImmutableAndFree() ?? originalConstraintTypes,
+                                                                      constraint.TypeConstraintsSyntax);
+                }
             }
 
-            if (clause1.ConstraintTypes.Length == 0 && clause2.ConstraintTypes.Length == 0)
+            if (builder != null)
             {
-                return true;
+                constraintClauses = builder.ToImmutableAndFree();
             }
 
-            var comparer = TypeWithAnnotations.EqualsComparer.Instance;
-            var constraintTypes1 = clause1.ConstraintTypes.ToImmutableHashSet(comparer);
-            var constraintTypes2 = clause2.ConstraintTypes.ToImmutableHashSet(comparer);
+            return constraintClauses;
 
-            foreach (var constraintType in constraintTypes1)
+            static bool mergeConstraints(ref TypeParameterConstraintKind mergedKind, ImmutableArray<TypeWithAnnotations> originalConstraintTypes,
+                                         ref SmallDictionary<TypeWithAnnotations, int> mergedConstraintTypesMap, ref ArrayBuilder<TypeWithAnnotations> mergedConstraintTypes,
```

Comment:
>mergedConstraintTypesMap [](start = 87, length = 24)

nit: not sure why this is called `*merged*ConstraintTypesMap`. Seems like `originalConstraintTypesMap` would be more accurate.